src/deluge/model/instrument/kit.cpp

```diff
@@ -1172,6 +1177,52 @@ void Kit::receivedNoteForDrum(ModelStackWithTimelineCounter* modelStack, MIDIDev
 	}
 }
 
+void Kit::possiblySetSelectedDrumAndRefreshUI(Drum* thisDrum) {
+	if (midiEngine.midiSelectKitRow) {
+		//is the drum for the note received the same as the currently selected drum in the kit?
+		//if no, update selected drum
+		if (selectedDrum != thisDrum) {
+			selectedDrum = thisDrum;
+
+			//is the instrument clip for this kit the currently selected clip? (could be selected by holding clip in song)
+			//only need to potentially refresh the kit clip if we're actively in that clip
+			//we use this function from midi follow because it was designed to detect clip selections in song / arranger view
+			//and when a clip is selected in song / arranger view, you will need to refresh the mod led indicators
+			//and if using midi follow you may need to send updated midi feedback (because parameter context has changed)
+			Clip* clip = midiFollow.getClipForMidiFollow();
+			if (clip && clip == activeClip) {
+				//are we currently in the instrument clip UI?
+				//if yes, we may need to refresh it (main pads and / or sidebar)
+				UI* currentUI = getCurrentUI();
+				if (currentUI == &instrumentClipView || currentUI == &automationInstrumentClipView) {
```

Comment:
Can you call instrumentClipView::setSelectedDrum instead? There's a bunch of duplicated logic here

Reply:
> Can you call instrumentClipView::setSelectedDrum instead? There's a bunch of duplicated logic here

the problem with using InstrumentClipView::setSelectedDrum is that it uses getCurrentKit() and getCurrentClip(), which means it won't work with pressing a clip in Song or Arranger because currentSong->currentClip isn't updated.

I did however make two improvements to setSelectedDrum because I noticed a couple things:

1) it would set the selected drum even if it's the same drum that's currently selected. This happens when you audition a kit row for instance. So I added a check to make sure we're not re-selecting the same drum. In that instance the only thing that would happen is a re-rendering of the audition pad (so it lights up when you press it).

2) it would set the mod controllable when changing the selected drum, even if affect entire is enabled. This did not make sense because with affect entire enabled you cannot control the mod controllable stack of the selected drum. So there is no need to redraw mod led's or set the mod controllable. So I added a check to see if affect entire is enabled

Reply:
> Can you call instrumentClipView::setSelectedDrum instead? There's a bunch of duplicated logic here

Just as a further follow up. I have further refactored setSelectedDrum which enables me to now call that function instead of adding duplicate code to the Kit class.